include/sampleflow/producers/differential_evaluation_mh.h

```diff
@@ -0,0 +1,201 @@
+         *   $x$, or using a Gaussian probability distribution centered at
+         *   $x$), then the ratio returned as second argument is
+         *   $\frac{\pi_\text{proposal}(\tilde x|x)}
+         *           {\pi_\text{proposal}(x|\tilde x)}=1$.
+         * @param[in] crossover A function that accepts the current sample as well
+         *   as two samples from other chains as arguments and combines them to
+         *   generate a new sample. This is the function that adds the
+         *   "differential evaluation" or "differential evolution" to Metropolis
+         *   Hastings. If the `OutputType` belongs to a vector space, the simple
+         *   solution is
+         *   ```
+         *   return current_sample + gamma * (sample_a - sample_b)
+         *   ```
+         *   where `gamma` is a scaling parameter; $\frac{2.38}{\sqrt{2d}}, d = $ the
+         *   dimension of the vector space, is commonly used.
+         * @param[in] n_chains The number of sample chains to produce in parallel.
+         * @param[in] crossover_gap The number of iterations in between crossover
+         *   iterations.
+         * @param[in] n_samples The number of (new) samples to be produced
+         *   by this function. This is also the number of times the
+         *   signal is called that notifies Consumer objects that a new
+         *   sample is available.
+         */
+        void
+        sample (const std::valarray<OutputType> starting_points,
+                const std::function<double (const OutputType &)> &log_likelihood,
+                const std::function<std::pair<OutputType,double> (const OutputType &)> &perturb,
+                const std::function<OutputType (const OutputType &, const OutputType &, const OutputType &)> &crossover,
+                const unsigned int crossover_gap,
+                const types::sample_index n_samples);
+    };
+
+
+    template <typename OutputType>
+    void
+    DifferentialEvaluationMetropolisHastings<OutputType>::
+    sample (const std::valarray<OutputType> starting_points,
+            const std::function<double (const OutputType &)> &log_likelihood,
+            const std::function<std::pair<OutputType,double> (const OutputType &)> &perturb,
+            const std::function<OutputType (const OutputType &, const OutputType &, const OutputType &)> &crossover,
+            const unsigned int crossover_gap,
+            const types::sample_index n_samples)
+    {
+      unsigned int n_chains = starting_points.size();
+      assert (n_chains >= 3);
+      // Make sure the flush_consumers() function is called at any point
+      // where we exit the current function.
+      Utilities::ScopeExit scope_exit ([this]()
+      {
+        this->flush_consumers();
+      });
+
+      std::mt19937 rng;
+      // Initialize distribution for comparing to acceptance ratio
+      std::uniform_real_distribution<> uniform_distribution(0,1);
+
+      std::valarray<OutputType> current_samples = starting_points;
```

Comment:
Use `std::vector` because there is no assumption that we can do arithmetic on `OutputType`.